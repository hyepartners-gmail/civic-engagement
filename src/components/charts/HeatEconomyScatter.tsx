"use client";

import { useMemo } from 'react';
import { 
  ScatterChart, 
  Scatter, 
  XAxis, 
  YAxis, 
  CartesianGrid, 
  Tooltip, 
  ResponsiveContainer,
  Line
} from 'recharts';

interface HeatEconomyScatterProps {
  points: [number, number, number][]; // [year, anomaly, proxy]
  colorByYear?: boolean;
  showFitLine?: boolean;
  fitLinePoints?: [number, number][]; // [anomaly, predicted_proxy]
  xAxisLabel?: string;
  yAxisLabel?: string;
}

export default function HeatEconomyScatter({ 
  points, 
  colorByYear = false,
  showFitLine = false,
  fitLinePoints = [],
  xAxisLabel = 'Summer Temperature Anomaly (°C)',
  yAxisLabel = 'Sector Productivity Proxy'
}: HeatEconomyScatterProps) {
  // Convert points to Recharts format
  const chartData = useMemo(() => {
    return points.map(([year, anomaly, proxy]) => ({
      x: anomaly,
      y: proxy,
      year
    }));
  }, [points]);
  
  // Calculate quartiles for color coding
  const yearQuartiles = useMemo(() => {
    if (!colorByYear || chartData.length === 0) return null;
    
    const years = chartData.map(d => d.year).sort((a, b) => a - b);
    const n = years.length;
    
    return {
      q1: years[Math.floor(n * 0.25)],
      q2: years[Math.floor(n * 0.5)],
      q3: years[Math.floor(n * 0.75)]
    };
  }, [chartData, colorByYear]);
  
  // Determine color for each point based on year
  const getColor = (year: number) => {
    if (!colorByYear || !yearQuartiles) return '#3b82f6'; // Default blue
    
    if (year <= yearQuartiles.q1) return '#3b82f6'; // Blue (earliest)
    if (year <= yearQuartiles.q2) return '#8b5cf6'; // Violet
    if (year <= yearQuartiles.q3) return '#f59e0b'; // Amber
    return '#ef4444'; // Red (latest)
  };
  
  if (chartData.length === 0) {
    return (
      <div className="h-80 flex items-center justify-center text-gray-500">
        <p>No data available for scatter plot</p>
      </div>
    );
  }

  return (
    <div className="h-80">
      <ResponsiveContainer width="100%" height="100%">
        <ScatterChart
          margin={{ top: 20, right: 20, bottom: 60, left: 60 }}
        >
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis 
            type="number" 
            dataKey="x" 
            name="Temperature Anomaly"
            tick={{ fontSize: 12 }}
            label={{ 
              value: xAxisLabel, 
              position: 'insideBottom', 
              offset: -40 
            }}
          />
          <YAxis 
            type="number" 
            dataKey="y" 
            name="Productivity Proxy"
            tick={{ fontSize: 12 }}
            label={{ 
              value: yAxisLabel, 
              angle: -90, 
              position: 'insideLeft' 
            }}
          />
          <Tooltip 
            formatter={(value, name) => {
              if (name === 'x') return [`${Number(value).toFixed(2)}°C`, 'Temp Anomaly'];
              if (name === 'y') return [Number(value).toFixed(2), 'Proxy Value'];
              return [String(value), ''];
            }}
            labelFormatter={() => ''}
            content={({ payload }: any) => {
              if (!payload || payload.length === 0) return null;
              const data = payload[0].payload;
              return (
                <div className="bg-white p-2 border rounded shadow">
                  <p className="font-semibold">Year: {data.year}</p>
                  <p>Temp Anomaly: {data.x.toFixed(2)}°C</p>
                  <p>Proxy Value: {data.y.toFixed(2)}</p>
                </div>
              );
            }}
          />
          {showFitLine && fitLinePoints.length > 0 && (
            <Line
              type="monotone"
              dataKey="y"
              data={fitLinePoints.map(([x, y]) => ({ x, y }))}
              stroke="#10b981"
              strokeWidth={2}
              dot={false}
              activeDot={false}
              legendType="none"
            />
          )}
          <Scatter 
            name="Data Points" 
            data={chartData} 
            fill="#3b82f6"
            shape={(props: any) => {
              const { cx, cy, payload } = props;
              return (
                <circle
                  cx={cx}
                  cy={cy}
                  r={6}
                  fill={getColor(payload.year)}
                  stroke="#fff"
                  strokeWidth={1}
                />
              );
            }}
          />
        </ScatterChart>
      </ResponsiveContainer>
    </div>
  );
}