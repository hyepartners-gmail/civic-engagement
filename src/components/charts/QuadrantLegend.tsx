"use client";

interface QuadrantLegendProps {
  xThreshold: number; // Temperature anomaly threshold (typically 0°C)
  yThreshold: number; // Productivity proxy threshold (typically median)
  xAxisLabel?: string;
  yAxisLabel?: string;
}

export default function QuadrantLegend({ 
  xThreshold, 
  yThreshold,
  xAxisLabel = 'Temperature Anomaly',
  yAxisLabel = 'Productivity Proxy'
}: QuadrantLegendProps) {
  const quadrants = [
    {
      id: 'hotter-worse',
      title: 'Hotter & Worse',
      description: `Above average ${xAxisLabel.toLowerCase()} and below average ${yAxisLabel.toLowerCase()}`,
      color: 'bg-red-100 border-red-300',
      position: 'top-right'
    },
    {
      id: 'hotter-better',
      title: 'Hotter & Better',
      description: `Above average ${xAxisLabel.toLowerCase()} and above average ${yAxisLabel.toLowerCase()}`,
      color: 'bg-orange-100 border-orange-300',
      position: 'bottom-right'
    },
    {
      id: 'cooler-worse',
      title: 'Cooler & Worse',
      description: `Below average ${xAxisLabel.toLowerCase()} and below average ${yAxisLabel.toLowerCase()}`,
      color: 'bg-blue-100 border-blue-300',
      position: 'top-left'
    },
    {
      id: 'cooler-better',
      title: 'Cooler & Better',
      description: `Below average ${xAxisLabel.toLowerCase()} and above average ${yAxisLabel.toLowerCase()}`,
      color: 'bg-green-100 border-green-300',
      position: 'bottom-left'
    }
  ];

  return (
    <div className="grid grid-cols-2 gap-4 mt-4">
      {quadrants.map((quadrant) => (
        <div 
          key={quadrant.id}
          className={`p-3 border rounded-lg ${quadrant.color}`}
        >
          <h4 className="font-semibold text-sm">{quadrant.title}</h4>
          <p className="text-xs text-gray-600 mt-1">{quadrant.description}</p>
          <div className="flex items-center mt-2 text-xs">
            <span className="bg-gray-200 px-1 py-0.5 rounded">
              {xAxisLabel}: {xThreshold >= 0 ? `≥ ${xThreshold}°C` : `≤ ${xThreshold}°C`}
            </span>
            <span className="mx-1">·</span>
            <span className="bg-gray-200 px-1 py-0.5 rounded">
              {yAxisLabel}: {quadrant.id.includes('worse') ? `< ${yThreshold.toFixed(1)}` : `≥ ${yThreshold.toFixed(1)}`}
            </span>
          </div>
        </div>
      ))}
    </div>
  );
}