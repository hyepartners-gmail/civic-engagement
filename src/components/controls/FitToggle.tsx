"use client";

import { useUrlState } from '@/hooks/useUrlState';
import { Label } from "@/components/ui/label";
import { Switch } from "@/components/ui/switch";

interface FitToggleProps {
  enabled: boolean;
  onToggle: (enabled: boolean) => void;
  r2?: number;
}

export default function FitToggle({ enabled, onToggle, r2 }: FitToggleProps) {
  const [urlFit, setUrlFit] = useUrlState<string>('fit', 'none');

  const handleToggle = (newState: boolean) => {
    onToggle(newState);
    setUrlFit(newState ? 'linear' : 'none');
  };

  return (
    <div className="flex items-center space-x-2">
      <Label htmlFor="fit-toggle">Show Trend Line</Label>
      <Switch
        id="fit-toggle"
        checked={enabled}
        onCheckedChange={handleToggle}
      />
      
      {enabled && r2 !== undefined && r2 > 0 && (
        <div className="text-sm text-gray-600">
          R² = {r2.toFixed(3)}
        </div>
      )}
      
      {enabled && r2 !== undefined && r2 === 0 && (
        <div className="text-sm text-yellow-600">
          Insufficient data for R²
        </div>
      )}
    </div>
  );
}