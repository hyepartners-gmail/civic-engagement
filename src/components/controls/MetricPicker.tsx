"use client";

import { useUrlState } from '@/hooks/useUrlState';
import { 
  Select, 
  SelectContent, 
  SelectItem, 
  SelectTrigger, 
  SelectValue 
} from '@/components/ui/select';

interface MetricPickerProps {
  metric: string;
  onMetricChange: (metric: string) => void;
}

const METRIC_OPTIONS = [
  { value: 'construction', label: 'Construction Hours' },
  { value: 'agriculture', label: 'Agriculture Yield Proxy' },
  { value: 'energy', label: 'Electric Load Proxy' }
];

export default function MetricPicker({ metric, onMetricChange }: MetricPickerProps) {
  const [urlMetric, setUrlMetric] = useUrlState<string>('metric', 'construction');

  const handleMetricChange = (newMetric: string) => {
    onMetricChange(newMetric);
    setUrlMetric(newMetric);
  };

  return (
    <div className="flex flex-col space-y-2">
      <label className="text-sm font-medium">Sector Metric</label>
      <Select value={metric} onValueChange={handleMetricChange}>
        <SelectTrigger className="w-48">
          <SelectValue placeholder="Select a metric" />
        </SelectTrigger>
        <SelectContent>
          {METRIC_OPTIONS.map((option) => (
            <SelectItem key={option.value} value={option.value}>
              {option.label}
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
      
      <div className="text-xs text-gray-500 mt-1">
        <p>Proxy metrics for economic sectors. Real data from BLS/BEA/USDA coming soon.</p>
      </div>
    </div>
  );
}