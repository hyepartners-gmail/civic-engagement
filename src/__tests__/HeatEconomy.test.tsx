import { render } from '@testing-library/react';
import HeatEconomyScatter from '@/components/charts/HeatEconomyScatter';
import QuadrantLegend from '@/components/charts/QuadrantLegend';
import MetricPicker from '@/components/controls/MetricPicker';
import FitToggle from '@/components/controls/FitToggle';

// Mock data for testing
const mockPoints: [number, number, number][] = [
  [2000, 0.5, 85],
  [2001, 0.6, 87],
  [2002, 0.7, 86],
  [2003, 0.8, 84],
  [2004, 0.9, 82]
];

const mockFitLinePoints: [number, number][] = [
  [0.5, 85],
  [0.9, 82]
];

describe('Heat Economy Components', () => {
  describe('HeatEconomyScatter', () => {
    it('should render correctly with data points', () => {
      const { container } = render(
        <HeatEconomyScatter 
          points={mockPoints} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with color by year', () => {
      const { container } = render(
        <HeatEconomyScatter 
          points={mockPoints} 
          colorByYear={true}
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with fit line', () => {
      const { container } = render(
        <HeatEconomyScatter 
          points={mockPoints} 
          showFitLine={true}
          fitLinePoints={mockFitLinePoints}
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render no data message when no points', () => {
      const { container } = render(
        <HeatEconomyScatter 
          points={[]} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('QuadrantLegend', () => {
    it('should render correctly with default labels', () => {
      const { container } = render(
        <QuadrantLegend 
          xThreshold={0}
          yThreshold={85}
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with custom labels', () => {
      const { container } = render(
        <QuadrantLegend 
          xThreshold={0}
          yThreshold={85}
          xAxisLabel="Temp Anomaly"
          yAxisLabel="Construction Hours"
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('MetricPicker', () => {
    it('should render correctly with construction metric', () => {
      const { container } = render(
        <MetricPicker 
          metric="construction" 
          onMetricChange={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with agriculture metric', () => {
      const { container } = render(
        <MetricPicker 
          metric="agriculture" 
          onMetricChange={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('FitToggle', () => {
    it('should render correctly when disabled', () => {
      const { container } = render(
        <FitToggle 
          enabled={false} 
          onToggle={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly when enabled without RÂ²', () => {
      const { container } = render(
        <FitToggle 
          enabled={true} 
          onToggle={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly when enabled with RÂ²', () => {
      const { container } = render(
        <FitToggle 
          enabled={true} 
          onToggle={jest.fn()} 
          r2={0.85}
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });
});