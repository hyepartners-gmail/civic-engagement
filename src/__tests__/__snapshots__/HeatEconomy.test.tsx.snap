// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heat Economy Components HeatEconomyScatter should render no data message when no points 1`] = `
<div>
  <div
    class="h-80 flex items-center justify-center text-gray-500"
  >
    <p>
      No data available for scatter plot
    </p>
  </div>
</div>
`;

exports[`Heat Economy Components QuadrantLegend should render correctly with custom labels 1`] = `
<div>
  <div
    class="grid grid-cols-2 gap-4 mt-4"
  >
    <div
      class="p-3 border rounded-lg bg-red-100 border-red-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Hotter & Worse
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Above average temp anomaly and below average construction hours
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temp Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Construction Hours
          : 
          &lt; 85.0
        </span>
      </div>
    </div>
    <div
      class="p-3 border rounded-lg bg-orange-100 border-orange-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Hotter & Better
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Above average temp anomaly and above average construction hours
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temp Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Construction Hours
          : 
          ≥ 85.0
        </span>
      </div>
    </div>
    <div
      class="p-3 border rounded-lg bg-blue-100 border-blue-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Cooler & Worse
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Below average temp anomaly and below average construction hours
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temp Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Construction Hours
          : 
          &lt; 85.0
        </span>
      </div>
    </div>
    <div
      class="p-3 border rounded-lg bg-green-100 border-green-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Cooler & Better
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Below average temp anomaly and above average construction hours
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temp Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Construction Hours
          : 
          ≥ 85.0
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Heat Economy Components QuadrantLegend should render correctly with default labels 1`] = `
<div>
  <div
    class="grid grid-cols-2 gap-4 mt-4"
  >
    <div
      class="p-3 border rounded-lg bg-red-100 border-red-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Hotter & Worse
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Above average temperature anomaly and below average productivity proxy
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temperature Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Productivity Proxy
          : 
          &lt; 85.0
        </span>
      </div>
    </div>
    <div
      class="p-3 border rounded-lg bg-orange-100 border-orange-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Hotter & Better
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Above average temperature anomaly and above average productivity proxy
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temperature Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Productivity Proxy
          : 
          ≥ 85.0
        </span>
      </div>
    </div>
    <div
      class="p-3 border rounded-lg bg-blue-100 border-blue-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Cooler & Worse
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Below average temperature anomaly and below average productivity proxy
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temperature Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Productivity Proxy
          : 
          &lt; 85.0
        </span>
      </div>
    </div>
    <div
      class="p-3 border rounded-lg bg-green-100 border-green-300"
    >
      <h4
        class="font-semibold text-sm"
      >
        Cooler & Better
      </h4>
      <p
        class="text-xs text-gray-600 mt-1"
      >
        Below average temperature anomaly and above average productivity proxy
      </p>
      <div
        class="flex items-center mt-2 text-xs"
      >
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Temperature Anomaly
          : 
          ≥ 0°C
        </span>
        <span
          class="mx-1"
        >
          ·
        </span>
        <span
          class="bg-gray-200 px-1 py-0.5 rounded"
        >
          Productivity Proxy
          : 
          ≥ 85.0
        </span>
      </div>
    </div>
  </div>
</div>
`;
